logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,173 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-                    addNeighbours(newVar, constraint.asOwns().attribute());
-                    constraint.asOwns().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isPlays()) {
-                    newVar.plays(constraint.asPlays().relation().flatMap(var -> Optional.of(convert(var))).orElse(null),
-                                 convert(constraint.asPlays().role()),
-                                 constraint.asPlays().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
-                    addNeighbours(newVar, constraint.asPlays().role());
-                    constraint.asPlays().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isRelates()) {
-                    newVar.relates(convert(constraint.asRelates().role()),
-                                   constraint.asRelates().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
-                    addNeighbours(newVar, constraint.asRelates().role());
-                    constraint.asRelates().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isIs()) {
-                    newVar.is(convert(constraint.asIs().variable()));
-                } else throw GraknException.of(ILLEGAL_STATE);
+            Set<ValueType> valueTypes = findComparableValueTypes(constraint);
+            if (valueTypeRegister.get(owner.id()).isEmpty()) {
+                valueTypes.forEach(valueType -> resolverTraversal.valueType(owner.id(), valueType));
+                valueTypeRegister.put(owner.id(), valueTypes);
+            } else if (!valueTypeRegister.get(owner.id()).containsAll(valueTypes)) {
+                throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
             }
-            return newVar;
+            subAttribute(owner.id());
         }
 
-    }
-
-    private static class Resolvers {
-
-        private final Map<Identifier, TypeVariable> typeResolvers;
-        private final Map<RelationConstraint.RolePlayer, TypeVariable> roleTypeResolvers;
-        // TODO: Why is there no checkers and getters for 'roleTypeResolvers'?
-        // TODO: do we still need 'roleTypeResolvers' given that we no longer put resolveRoleTypes on RelationConstraints?
-
-        private int sysVarCounter;
-
-        Resolvers() {
-            this.sysVarCounter = 0;
-            this.typeResolvers = new HashMap<>();
-            this.roleTypeResolvers = new HashMap<>();
-        }
-
-        TypeVariable register(RelationConstraint.RolePlayer rolePlayer) {
-            return roleTypeResolvers.computeIfAbsent(rolePlayer, rp -> new TypeVariable(newSystemVariable()));
+        private Set<ValueType> findComparableValueTypes(ValueConstraint<?> constraint) {
+            if (constraint.isBoolean()) return set(ValueType.BOOLEAN);
```

Comment:
I think you can replace lots of this by getting it from: `Encoding.ValueType.of(constraint).comparables()`!

Reply:
It's a nice idea, but it ends up ugly. You cannot call `Encoding.ValueType()` with a `constraint`. You need to use `constraint.value()`. But you have to exclude the possibility that it's a Variable Value Constraint, which leaves us with something like:

```
if (!constraint.isVariable()) {
                Set<Encoding.ValueType> comparables = Encoding.ValueType.of((ValueType) constraint.value()).comparables();
                return comparables.stream().map(Encoding.ValueType::graqlValueType).collect(Collectors.toSet());
            }
```

Which I think is far less clear than:
```
            if (constraint.isBoolean()) return set(ValueType.BOOLEAN);
            else if (constraint.isString()) return set(ValueType.STRING);
            else if (constraint.isDateTime()) return set(ValueType.DATETIME);
            else if (constraint.isLong() || constraint.isDouble()) return set(ValueType.LONG, 
```

However, I do think the logic of what's comparable to what should exist elsewhere, since that's something that will be universal to the codebase, and in the future there will be a lot more combinations of value types that are comparable... so we should return to this question later.

Reply:
I think we should not keep the code as you've written for precisely the duplication you've mentioned, the more times we duplicate whats "comparable", the more we can shoot ourselves in the foot. I think we should trade off some readability for a more standardised comparison set :)

Reply:
I agree to avoid duplication, but I'm no sure that `Encoding.ValueType` is in a great format for dealing with that. 

I will use it as it is for now, but I suspect we will want to refactor it as it comes again.